### PR TITLE
to_int can round numeric values

### PR DESCRIPTION
**Pull request by DelazJ**
Also update the tags with more pertinent references

## Description

[Replace this with some text explaining the rationale and details about this pull request]

### CODE REVIEW

File: resources/function_help/json/to_int [hunk 1]
```diff
@@ -2,14 +2,20 @@
   "name": "to_int",
   "type": "function",
   "groups": ["Conversions"],
-  "description": "Converts a string to integer number. Nothing is returned if a value cannot be converted to integer (e.g '123asd' is invalid).",
+  "description": "Rounds a numeric value to integer or converts a string to integer number. Nothing is returned if the value cannot be converted to integer (e.g '123asd' or '123.45' is invalid).",
   "arguments": [{
```
Comment (nyalldawson) [inline resources/function_help/json/to_int:6]: Why is 123.45 invalid?

Comment (DelazJ) [reply resources/function_help/json/to_int:6]: it is '123.45', as string. Why invalid? I can't explain, I just observed. maybe a left-over?

image

Comment (nyalldawson) [reply resources/function_help/json/to_int:6]: Hmm, that's very confusing. The function can convert a string to number, OR a double to into BUT not both?

Comment (agiudiceandrea) [reply resources/function_help/json/to_int:6]: It seems a Qt feature/issue (`to_int `uses `QgsExpressionUtils::getIntValue` which uses `QVariant.toLongLong`)...

```Python
val = QVariant(10)
val.convert(QMetaType.LongLong)
True
val.value()
10

val = QVariant('10')
val.convert(QMetaType.LongLong)
True
val.value()
10

val = QVariant(10.6)
val.convert(QMetaType.LongLong)
True
val.value()
11

val = QVariant('10.6')
val.convert(QMetaType.LongLong)
False
val.value()
NULL

```

Comment (roya0045) [reply resources/function_help/json/to_int:6]: Would it be safe to check for decimal separator, split the string, int the first half and increase if the bottom is lower? Or do a string to float to int?

Comment (nyalldawson) [reply resources/function_help/json/to_int:6]: > Or do a string to float to int?

That sounds right 👍

Comment (DelazJ) [reply resources/function_help/json/to_int:6]: That's again beyond my skills. I welcome a commit or a replacement PR. Sorry.

Comment (roya0045) [reply resources/function_help/json/to_int:6]: TBH I'm not sure if the safety would be better in `QgsExpressionUtils::getIntValue` or only in `fcnToInt`, as the first is used in many expressions internally, but I'm not sure if the added safety is needed.

**Comment by roya0045** [comment]
@DelazJ I made a blind change just as an example, hopefully the following link will work, https://github.com/qgis/QGIS/commit/ab5f4fc10aa74e90a202c6eece71401097d82e79

If you want you can merge me branch into yours.

**Comment by DelazJ** [comment]
Thanks @roya0045 I confirm it works using an English locale (the screenshot is taken from French locale but it was OK on English).
image
I tried with French locale where comma is supposed to be the decimal character and it does not work. But I can' tell what are really the settings in the Options tab for French due to #50239 
image

**Comment by roya0045** [comment]
@DelazJ this might fix it (also handle double should be better), though the function getDouble doesn't convert to handle , separators, You could test the other conversions to see how it goes, 

I could have used the function getDoubleValue internally also, but it would fix the conversion but might be easier to maintain. Would require the QLocale method also.

**Comment by nicogodet** [comment]
@DelazJ I think QGIS expression never worked with comma as decimal separator. It is not specific to this PR
`to_real('123,45')` raises an error.

**Comment by roya0045** [comment]
> @DelazJ I think QGIS expression never worked with comma as decimal separator. It is not specific to this PR `to_real('123,45')` raises an error.

I can rework my branch and it could fix both I think, as mentionned in my previous post.

**Comment by roya0045** [comment]
@DelazJ @nicogodet are we sure that commas will always be used? If so https://github.com/qgis/QGIS/commit/d01075431182127cbe5ebaca2d1e6686a68c1fc1 should be safe, but I'm not sure if it would be better to check the local first, try the comma variant, and then still test for a period with the more standard way of doing decimal, just as a fallback or a default. 

All of that could be move to another function to prevent duplicate code.

**Comment by roya0045** [comment]
Forgot a type change in one of the previous commit, had 5 min of free time, did the internal handler if someone wanted to test: https://github.com/qgis/QGIS/commit/012f68c27295e8470cb6d4748386f98d64a830fc

**Comment by DelazJ** [comment]
@roya0045 sorry for the late reply. I'm afraid I don't understand what you meant by
> are we sure that commas will always be used?

Anyway, what do you need to move this forward?

**Comment by roya0045** [comment]
> @roya0045 sorry for the late reply. I'm afraid I don't understand what you meant by
> 
> > are we sure that commas will always be used?
> 
> Anyway, what do you need to move this forward?

No  worries, if you want you can try the last commit I mentionned  (https://github.com/qgis/QGIS/commit/012f68c27295e8470cb6d4748386f98d64a830fc ) as that should be somewhat more reliable as it tries a standard conversion and then check with local handling. Another option would have been to convert the comma to a point, but that felt more risky.

**Pull request closed without merging**